src/ngraph/pass/constant_folding_non_zero.cpp

```diff
@@ -0,0 +1,198 @@
+
+        // we can't create an empty Constant indicating that no non-zero elems were found
+        NGRAPH_CHECK(
+            found_indices.front().size() > 0,
+            "It's not possible to constant fold a NonZero op with an input containing only zeros.");
+
+        // flatten the results and return them as a Constant
+        std::vector<int64_t> indices;
+        indices.reserve(found_indices.size() * found_indices.front().size());
+        for (const auto& row : found_indices)
+        {
+            indices.insert(indices.end(), row.begin(), row.end());
+        }
+
+        const Shape out_shape{found_indices.size(), found_indices.front().size()};
+        return op::Constant::create(element::i64, out_shape, indices);
+    }
+}
+
+void pass::ConstantFolding::construct_constant_non_zero()
+{
+    const auto data_label = make_shared<pattern::op::Label>(
+        element::f32, Shape{2, 2, 3}, pattern::has_class<op::Constant>());
+    const auto non_zero = make_shared<op::v3::NonZero>(data_label);
+
+    auto constant_non_zero_callback = [data_label](pattern::Matcher& m) {
+        auto pattern_map = m.get_pattern_map();
+
+        const auto data = static_pointer_cast<op::Constant>(pattern_map[data_label]);
+
+        // const auto found_nz_node = m.get_match_root();
+        // this fails because the output of NonZero is still dynamic - is it ok to remove this line?
+        // or should the output shape of this op be calculated as the maximum possible number
+        // of non-zero indices it can return if the input is a constant?
+        // NGRAPH_CHECK(revalidate_and_ensure_static(found_nz_node));
+
+        std::shared_ptr<Node> replacement;
+        switch (data->get_element_type())
+        {
+        case element::Type_t::bf16: replacement = fold_constant_non_zero<bfloat16>(data); break;
+        case element::Type_t::f16: replacement = fold_constant_non_zero<float16>(data); break;
+        case element::Type_t::f32: replacement = fold_constant_non_zero<float>(data); break;
+        case element::Type_t::f64: replacement = fold_constant_non_zero<double>(data); break;
+        case element::Type_t::i8: replacement = fold_constant_non_zero<int8_t>(data); break;
+        case element::Type_t::i16: replacement = fold_constant_non_zero<int16_t>(data); break;
+        case element::Type_t::i32: replacement = fold_constant_non_zero<int32_t>(data); break;
+        case element::Type_t::i64: replacement = fold_constant_non_zero<int64_t>(data); break;
+        case element::Type_t::u8: replacement = fold_constant_non_zero<uint8_t>(data); break;
+        case element::Type_t::u16: replacement = fold_constant_non_zero<uint16_t>(data); break;
+        case element::Type_t::u32: replacement = fold_constant_non_zero<uint32_t>(data); break;
+        case element::Type_t::u64: replacement = fold_constant_non_zero<uint64_t>(data); break;
+        case element::Type_t::u1:
+        case element::Type_t::boolean:
```

Comment:
Just wonder if NonZero operator could return `indicies` of all true values for boolean type. It could be helpful in some cases.

Reply:
The spec only mentions numerical types for the input data, this is why I disabled bool in here.

Reply:
Let's add support for boolean data types also.